File: movies/app/src/main/java/com/callor/movies/MainActivity.java
package com.callor.movies;

import android.os.Bundle;

import com.google.android.material.snackbar.Snackbar;

import androidx.appcompat.app.AppCompatActivity;

import android.view.View;

import androidx.navigation.NavController;
import androidx.navigation.Navigation;
import androidx.navigation.ui.AppBarConfiguration;
import androidx.navigation.ui.NavigationUI;

import com.callor.movies.databinding.ActivityMainBinding;

import android.view.Menu;
import android.view.MenuItem;
import android.widget.Toast;

public class MainActivity extends AppCompatActivity {

    private AppBarConfiguration appBarConfiguration;
    private ActivityMainBinding binding;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        binding = ActivityMainBinding.inflate(getLayoutInflater());
        setContentView(binding.getRoot());

        setSupportActionBar(binding.toolbar);

//        content_main에 존재 : nav_host_fragment_content_main
        /*
        content_view.xml에 설정된 fragment view를
        NavController로 등록하여
        fragment 간의 이동을 쉽게 하겠다.
         */
        NavController navController
                = Navigation.findNavController(this, R.id.nav_host_fragment_content_main);
//        뒤로가기 버튼 등이 있을 때 처리를 쉽게 하기 위한 설정
//         ui에 뒤로가기 버튼이 생겼을 때 처리한다.
        appBarConfiguration = new AppBarConfiguration.Builder(navController.getGraph()).build();
        NavigationUI.setupActionBarWithNavController(this, navController, appBarConfiguration);

        binding.fab.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                Snackbar.make(view, "Replace with your own action", Snackbar.LENGTH_LONG)
                        .setAction("Action", null).show();
            }
        });
    }
/*
ActionBar에 메뉴 등을 표현하기 위하여 작성되는 코드
res/menu/menu_main.xml 파일을 읽어서
ActionBar에 메뉴를 그리는 코드가 작성된다.
 */
    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        // Inflate the menu; this adds items to the action bar if it is present.
        getMenuInflater().inflate(R.menu.menu_main, menu);
//        R.menu.menu_main불러와서 menu에 넣어라
//       여길 수정하려면 menu를 수정하면 된다.

        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        int id = item.getItemId();

        //noinspection SimplifiableIfStatement
        if (id == R.id.action_settings) {

            Toast.makeText(MainActivity.this, "설정메뉴클릭"
            ,Toast.LENGTH_LONG).show();
            return true;
        }

        return super.onOptionsItemSelected(item);
    }
/*
안드로이드의 물리적 버튼
원래 어플에 존재하는 버튼들을 의미(홈, 뒤로가기, 백그라운드목록)
위로가기,뒤로가기,어플종료,어플리스트 보기 등의 기능을 수행하는 활성화 시키는 event 핸들러

안드로이드 Up 버튼을 눌렀을 때 Navigation에 설정된 대로
뒤로가기를 수행하다가 가장 먼저 열린 화면에 도달했을 때
Up버튼을 누르면 어플을 종료하는 코드가 자동으로 실행된다.
 */
    @Override
    public boolean onSupportNavigateUp() {
        NavController navController =
                Navigation.findNavController(this, R.id.nav_host_fragment_content_main);
        return NavigationUI.navigateUp(navController, appBarConfiguration)
                || super.onSupportNavigateUp();
    }
}

File: HelloApp2/app/src/main/java/com/callor/hello/MainActivity.kt
package com.callor.hello

import android.os.Bundle
import com.google.android.material.snackbar.Snackbar
import androidx.appcompat.app.AppCompatActivity
import androidx.navigation.findNavController
import androidx.navigation.ui.AppBarConfiguration
import androidx.navigation.ui.navigateUp
import androidx.navigation.ui.setupActionBarWithNavController
import android.view.Menu
import android.view.MenuItem
import com.callor.hello.databinding.ActivityMainBinding

class MainActivity : AppCompatActivity() {
    /**
     * NoActionBar Theme 화면에서
     * Custom 한 AcitonBar(title bar)를 생성하기 위한 클래스
     */

    private lateinit var appBarConfiguration: AppBarConfiguration
//    res/layout/activity_main.xml 파일을 클래스코에서 핸들링을 할 때는
//    xml 파일을 사용하여 자동으로 클래스가 생성된다.
//    xml 파일을 만들어놓으면 이렇게 클래스로 사용할 수 있게 한다.
//    사용되게 하려면 xml 이 클래스로 자동으로 변경된다 신기방기
    private lateinit var binding: ActivityMainBinding

//    Android App 에서 화면 View 와 관련된 클래스
//    main() 진입점 클래스
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

    /**
     * activity_main.xml 의 view Component 를 한꺼번에 바인딩 하기 위한 코드
     * 예) 과거의 Java 코드에서 xml 에 포함된 view Component 들을 사용하기 위하여 초기화  다음과 같은 코드를
     * 사용하지 않는다.
     * 한번 바인딩 해주면 밑에 코드처럼 사용하지 않고 바로 사용할 수 있게 된다.
     * TextEdit textEdit = findById(R.Id.txt_01)
     */
        binding = ActivityMainBinding.inflate(layoutInflater)
    /**
     * activity_main.xml 전체를 확장하여 처음 어플 화면에 보여라
     * inflate를 하게 되면 root 라는 것이 생긴다.
     * 이 root는 컴포넌트 하나하나를 모아놓은 전체를 갖고 있다.
     */
        setContentView(binding.root)
//  임의로 생성한 ActionBar를 현재 화면에 같이 부착하라
//   id가 Toolbar 인 ActionBar 컴포넌트를 현재 화며에 부착하라.
        setSupportActionBar(binding.toolbar)

        val navController = findNavController(R.id.nav_host_fragment_content_main)
        appBarConfiguration = AppBarConfiguration(navController.graph)
        setupActionBarWithNavController(navController, appBarConfiguration)

        binding.fab.setOnClickListener { view ->
            Snackbar.make(view, "Replace with your own action", Snackbar.LENGTH_LONG)
                    .setAction("Action", null).show()
        }
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        // Inflate the menu; this adds items to the action bar if it is present.
        menuInflater.inflate(R.menu.menu_main, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        // Handle action bar item clicks here. The action bar will
        // automatically handle clicks on the Home/Up button, so long
        // as you specify a parent activity in AndroidManifest.xml.
        return when (item.itemId) {
            R.id.action_settings -> true
            else -> super.onOptionsItemSelected(item)
        }
    }

    override fun onSupportNavigateUp(): Boolean {
        val navController = findNavController(R.id.nav_host_fragment_content_main)
        return navController.navigateUp(appBarConfiguration)
                || super.onSupportNavigateUp()
    }
}